packages/protobuf/src/next/to-binary.ts

```diff
@@ -0,0 +1,236 @@
+  (writer.tag(fieldNo, wireType)[method] as (v: unknown) => void)(value);
+}
+
+function writeMessageField(
+  writer: IBinaryWriter,
+  opts: BinaryWriteOptions,
+  field: DescField &
+    ({ fieldKind: "message" } | { fieldKind: "list"; listKind: "message" }),
+  message: Message,
+) {
+  const rm = reflectToBinary(reflect(message), opts);
+  if (
+    field.proto.type === FieldDescriptorProto_Type.GROUP ||
+    field.getFeatures().messageEncoding === FeatureSet_MessageEncoding.DELIMITED
+  ) {
+    writer
+      .tag(field.number, WireType.StartGroup)
+      .raw(rm)
+      .tag(field.number, WireType.EndGroup);
+  } else {
+    writer.tag(field.number, WireType.LengthDelimited).bytes(rm);
+  }
+}
+
+function writeListField(
+  writer: IBinaryWriter,
+  opts: BinaryWriteOptions,
+  field: DescField & { fieldKind: "list" },
+  value: readonly unknown[],
+) {
+  if (field.listKind == "message") {
+    for (const item of value) {
+      writeMessageField(writer, opts, field, item as Message);
+    }
+    return;
+  }
+  if (field.packed) {
+    if (!value.length) {
+      return;
+    }
+    writer.tag(field.number, WireType.LengthDelimited).fork();
+    const [, method] = scalarTypeInfo(field.scalar ?? ScalarType.INT32);
+    for (let i = 0; i < value.length; i++) {
+      (writer[method] as (v: unknown) => void)(value[i]);
+    }
+    writer.join();
+    return;
+  }
+  for (const item of value) {
+    writeScalar(writer, field.scalar ?? ScalarType.INT32, field.number, item);
+  }
+}
+
+function writeMapEntry(
+  writer: IBinaryWriter,
+  opts: BinaryWriteOptions,
+  field: DescField & { fieldKind: "map" },
+  key: string,
+  value: unknown,
+) {
+  writer.tag(field.number, WireType.LengthDelimited);
+  writer.fork();
+  // javascript only allows number or string for object properties
+  // we convert from our representation to the protobuf type
+  let keyValue: ScalarValue;
+  switch (field.mapKey) {
+    case ScalarType.INT32:
+    case ScalarType.FIXED32:
+    case ScalarType.UINT32:
+    case ScalarType.SFIXED32:
+    case ScalarType.SINT32:
+      keyValue = Number.parseInt(key);
+      break;
+    case ScalarType.BOOL:
+      keyValue = key == "true";
+      break;
+    default:
+      keyValue = key;
+      break;
+  }
+
+  // write key, expecting key field number = 1
+  writeScalar(writer, field.mapKey, 1, keyValue);
+
+  // write value, expecting value field number = 2
+  switch (field.mapKind) {
+    case "scalar":
+    case "enum":
+      writeScalar(writer, field.scalar ?? ScalarType.INT32, 2, value);
+      break;
+    case "message":
+      writer
+        .tag(2, WireType.LengthDelimited)
+        .bytes(reflectToBinary(reflect(value as Message), opts));
+      break;
+  }
+
+  writer.join();
+}
+
+/**
+ * Get information for writing a scalar value.
+ *
+ * Returns tuple:
+ * [0]: appropriate WireType
+ * [1]: name of the appropriate method of IBinaryWriter
+ */
+function scalarTypeInfo(
```

Comment:
Calling the writer's method with the bracket notation is awfully slow.  The v1 has a TODO comment to remove this function. 

Can you add it here as well, or do the change right away?